src/index.ts

```diff
@@ -93,142 +91,136 @@ function createExtensions(document, result, requestID) {
-      ...(INTROSPECT_TOKEN &&
-      req.headers["authorization"] !== `Bearer ${INTROSPECT_TOKEN}`
-        ? [NoSchemaIntrospectionCustomRule]
-        : []),
-    ]
-    if (QUERY_DEPTH_LIMIT) validationRules.push(depthLimit(QUERY_DEPTH_LIMIT))
-
-    return {
-      schema: appSchema,
-      graphiql: !PRODUCTION_ENV,
-      context,
-      rootValue: {},
-      customFormatErrorFn: graphqlErrorHandler(enableSentry, {
-        req,
-        // Why the checking on params? Do we reach this code if params is falsy?
-        variables: params && params.variables,
-        query: (params && params.query)!,
-      }),
-      validationRules,
-      extensions: ({ document, result }) =>
-        createExtensions(document, result, requestID),
-    }
+  const loaders = createLoaders(accessToken, userID, {
+    requestIDs,
+    userAgent,
+    appToken,
   })
 
-  app.use("/batch", bodyParser.json(), graphqlBatchHTTPWrapper(graphqlServer))
+  const exchangeSchema = executableExchangeSchema(legacyTransformsForExchange)
 
-  if (ENABLE_GRAPHQL_UPLOAD) {
-    app.use(
-      graphqlUploadExpress({
-        maxFileSize: GRAPHQL_UPLOAD_MAX_FILE_SIZE_IN_BYTES,
-        maxFiles: GRAPHQL_UPLOAD_MAX_FILES,
-      })
-    )
+  const context: ResolverContext = {
+    accessToken,
+    userID,
+    defaultTimezone,
+    ...loaders,
+    // For stitching purposes
+    exchangeSchema,
+    requestIDs,
+    userAgent,
+    appToken,
   }
 
-  app.use(graphqlServer)
-
-  if (enableSentry) {
-    app.use(Sentry.Handlers.errorHandler())
+  const validationRules = [
+    principalFieldDirectiveValidation,
+
+    // require Authorization header for introspection (in production if configured)
+    ...(INTROSPECT_TOKEN &&
+    req.headers["authorization"] !== `Bearer ${INTROSPECT_TOKEN}`
+      ? [NoSchemaIntrospectionCustomRule]
+      : []),
+  ]
+  if (QUERY_DEPTH_LIMIT) validationRules.push(depthLimit(QUERY_DEPTH_LIMIT))
+
+  return {
+    schema,
+    graphiql: !PRODUCTION_ENV,
+    context,
+    rootValue: {},
+    customFormatErrorFn: graphqlErrorHandler(enableSentry, {
+      req,
+      // Why the checking on params? Do we reach this code if params is falsy?
+      variables: params && params.variables,
+      query: (params && params.query)!,
+    }),
+    validationRules,
+    extensions: ({ document, result }) =>
+      createExtensions(document, result, requestID),
   }
+})
+
+app.use("/batch", bodyParser.json(), graphqlBatchHTTPWrapper(graphqlServer))
 
-  return app
+if (ENABLE_GRAPHQL_UPLOAD) {
+  app.use(
+    graphqlUploadExpress({
+      maxFileSize: GRAPHQL_UPLOAD_MAX_FILE_SIZE_IN_BYTES,
+      maxFiles: GRAPHQL_UPLOAD_MAX_FILES,
+    })
+  )
 }
 
-const app = express()
+app.use("/v2", (req, res, next) => {
+  if (req.url !== "/") return next()
+  return graphqlServer(req, res, next)
+})
 
-// This order is important for dd-trace to be able to find the nested routes.
-app.use("/v2", startApp(schemaV2, "/"))
-app.use("/", startApp(schemaV1, "/"))
```

Comment:
It would be nice if formerly-valid requests to the v1 endpoint received a meaningful error message rather than a nonspecific 404. (We _know_ there are continuing requests to this endpoint, either from the wild or from our own forgotten clients.)

The status code should still be `404`, but just something clarifying like `The root (v1) graphQL endpoint has been deprecated and clients should instead use the /v2 endpoint.`

Reply:
Since technically all random paths like "/blah/whatever" would match V1 , it almost sounds like you are suggesting to have a catch-all `app.use("*", ...` _after_ everything else, and have that return a 404 (as it would now), but w/ that custom message.

I don't think we have to be particularly precious about that kind of thing...what do you think?

Reply:
I don't think we have to support all _possible_ paths to v1, but a meaningful replacement for the official one would be nice.